app/features/transactions/transaction-hooks.ts

```diff
@@ -0,0 +1,97 @@
+import type { RealtimePostgresChangesPayload } from '@supabase/supabase-js';
+import { useQuery, useQueryClient } from '@tanstack/react-query';
+import { useEffect } from 'react';
+import {
+  type BoardwalkDbTransaction,
+  boardwalkDb,
+} from '~/features/boardwalk-db/database';
+import { useLatest } from '~/lib/use-latest';
+import type { Transaction } from './transaction';
+import {
+  TransactionRepository,
+  useTransactionRepository,
+} from './transaction-repository';
+
+export function useTransaction({
+  transactionId,
+  onCompleted,
+  onFailed,
+}: {
+  transactionId?: string;
+  onCompleted?: () => void;
+  onFailed?: () => void;
+}) {
+  const enabled = !!transactionId;
+  const onCompletedRef = useLatest(onCompleted);
+  const onFailedRef = useLatest(onFailed);
+  const queryClient = useQueryClient();
+  const transactionRepository = useTransactionRepository();
+
+  const { data: transaction } = useQuery({
+    queryKey: ['transaction', transactionId],
+    queryFn: () => transactionRepository.get(transactionId ?? ''),
+    enabled,
+    staleTime: Number.POSITIVE_INFINITY,
+  });
+
+  useOnTransactionChange({
+    transactionId,
+    onUpdated: (transaction) => {
+      queryClient.setQueryData<Transaction>(
+        ['transaction', transactionId],
+        transaction,
+      );
+    },
+  });
+
+  useEffect(() => {
```

Comment:
this would trigger onCompleted callback for the transaction tha was completed who knows when in the past right?

I wonder if we should just execute this logic from inside the useOnTransactionChange onUpdated